modules/ref_fb_module/src/classifier_fb_impl.cpp

```diff
@@ -0,0 +1,360 @@
+
+        if (outputName.empty())
+            outputDataDescriptorBuilder.setName(inputDataDescriptor.getName().toStdString() + "/Classifired");
+        else
+            outputDataDescriptorBuilder.setName(outputName);
+
+        outputDataDescriptorBuilder.setUnit(Unit("%"));
+
+        outputDataDescriptor = outputDataDescriptorBuilder.build();
+        outputSignal.setDescriptor(outputDataDescriptor);
+        
+        outputDomainDataDescriptor = DataDescriptorBuilderCopy(inputDomainDataDescriptor).setRule(ExplicitDataRule()).build();
+        outputDomainSignal.setDescriptor(outputDomainDataDescriptor);
+    }
+    catch (const std::exception& e)
+    {
+        LOG_W("Failed to set descriptor for power signal: {}", e.what())
+        outputSignal.setDescriptor(nullptr);
+    }
+}
+
+
+void ClassifierFbImpl::onPacketReceived(const InputPortPtr& port)
+{
+    std::scoped_lock lock(sync);
+
+    PacketPtr packet;
+    const auto connection = inputPort.getConnection();
+    if (!connection.assigned())
+        return;
+
+    packet = connection.dequeue();
+
+    while (packet.assigned())
+    {
+        switch (packet.getType())
+        {
+            case PacketType::Event:
+                processEventPacket(packet);
+                break;
+
+            case PacketType::Data:
+                SAMPLE_TYPE_DISPATCH(inputSampleType, processDataPacket, packet);
+                break;
+
+            default:
+                break;
+        }
+
+        packet = connection.dequeue();
+    };
+}
+
+void ClassifierFbImpl::processEventPacket(const EventPacketPtr& packet)
+{
+    if (packet.getEventId() == event_packet_id::DATA_DESCRIPTOR_CHANGED)
+    {
+        DataDescriptorPtr inputDataDescriptor = packet.getParameters().get(event_packet_param::DATA_DESCRIPTOR);
+        DataDescriptorPtr inputDomainDataDescriptor = packet.getParameters().get(event_packet_param::DOMAIN_DATA_DESCRIPTOR);
+        processSignalDescriptorChanged(inputDataDescriptor, inputDomainDataDescriptor);
+    }
+}
+
+inline bool ClassifierFbImpl::timeInInterval(UInt startTime, UInt endTime) {
+    return (endTime - startTime) < timeMs(blockSize);
+}
+
+template <SampleType InputSampleType>
+void ClassifierFbImpl::processDataPacket(const DataPacketPtr& packet)
+{
+    using InputType = typename SampleTypeToType<InputSampleType>::Type;
+    using OutputType = Float;
+
+    // if packet does not have samples - ignore
+    if (packet.getSampleCount() == 0) {
+        return;
+    }
+
+    packets.push_back(packet);
+
+    size_t outputPackages = 0;
+    {
+        auto inputDomainData = static_cast<UInt*>(packet.getDomainPacket().getData());
+        UInt lastTime = inputDomainData[packet.getSampleCount() - 1];
+
+        // initialize members
+        if (packetStarted == UInt{}) {
+            packetStarted = inputDomainData[0];
+            sampleStarted = 0; 
+        }
+
+        outputPackages = (lastTime - packetStarted) / timeMs(blockSize);
```

Comment:
The time is usually in ticks, not yet scaled to the unit of the signal. This works only if the unit of the signal is "seconds", and the resolution is 1/1000.